topic-operator/src/main/java/io/strimzi/operator/topic/TopicStoreTopologyProvider.java

```diff
@@ -0,0 +1,132 @@
+    ) {
+        this.storeTopic = storeTopic;
+        this.topicStoreName = topicStoreName;
+        this.kafkaProperties = kafkaProperties;
+        this.dispatcher = dispatcher;
+    }
+
+    @Override
+    public Topology get() {
+        StreamsBuilder builder = new StreamsBuilder();
+
+        // Simple defaults
+        Map<String, String> configuration = new HashMap<>();
+        configuration.put(TopicConfig.CLEANUP_POLICY_CONFIG, TopicConfig.CLEANUP_POLICY_COMPACT);
+        configuration.put(TopicConfig.MIN_COMPACTION_LAG_MS_CONFIG, "0");
+        configuration.put(TopicConfig.SEGMENT_BYTES_CONFIG, String.valueOf(64 * 1024 * 1024));
+
+        // Input topic command -- store topic
+        // Key is topic name -- which is also used for KeyValue store key
+        KStream<String, TopicCommand> topicRequest = builder.stream(
+                storeTopic,
+                Consumed.with(Serdes.String(), new TopicCommandSerde())
+        );
+
+        // Data structure holds all topic information
+        StoreBuilder<KeyValueStore<String /* topic */, Topic>> topicStoreBuilder =
+                Stores
+                        .keyValueStoreBuilder(
+                                Stores.inMemoryKeyValueStore(topicStoreName),
+                                Serdes.String(), new TopicSerde()
+                        )
+                        .withCachingEnabled()
+                        .withLoggingEnabled(configuration);
+
+        builder.addStateStore(topicStoreBuilder);
+
+        topicRequest.process(
+            () -> new TopicCommandTransformer(topicStoreName, dispatcher),
+            topicStoreName
+        );
+
+        return builder.build(kafkaProperties);
+    }
+
+    private static class TopicCommandTransformer implements Processor<String, TopicCommand> {
+        private final String topicStoreName;
+        private final ForeachAction<? super String, ? super Integer> dispatcher;
+
+        private KeyValueStore<String, Topic> store;
+
+        public TopicCommandTransformer(
+                String topicStoreName,
+                ForeachAction<? super String, ? super Integer> dispatcher
+        ) {
+            this.topicStoreName = topicStoreName;
+            this.dispatcher = dispatcher;
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public void init(ProcessorContext context) {
+            store = (KeyValueStore<String, Topic>) context.getStateStore(topicStoreName);
+        }
+
+        @Override
+        public void process(String key, TopicCommand value) {
+            String uuid = value.getUuid();
+            TopicCommand.Type type = value.getType();
+            Integer result = null;
+            switch (type) {
+                case CREATE:
+                    Topic previous = store.putIfAbsent(key, value.getTopic());
+                    if (previous != null) {
+                        result = KafkaStreamsTopicStore.toIndex(TopicStore.EntityExistsException.class);
```

Comment:
When we throw EEE like this should be have updated the store by side-effect? I assume not, since I think ZK would throw without updating the stored state.

Reply:
That's why I used putIfAbsent

Reply:
Ah, Mondays!